sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxDocumentClientImpl.java

```diff
@@ -1812,6 +1830,75 @@ public QueryCompatibilityMode getQueryCompatibilityMode() {
+        String collectionLink,
+        PartitionKey partitionKey,
+        CosmosQueryRequestOptions options) {
+
+        if (StringUtils.isEmpty(collectionLink)) {
+            throw new IllegalArgumentException("collectionLink");
+        }
+
+        if (partitionKey == null) {
+            throw new IllegalArgumentException("partitionKey");
+        }
+
+        RxDocumentServiceRequest request = RxDocumentServiceRequest.create(
+            OperationType.Query,
+            ResourceType.Document,
+            collectionLink,
+            null
+        );
+
+        // This should not got to backend
+        Flux<Utils.ValueHolder<DocumentCollection>> collectionObs =
+            collectionCache.resolveCollectionAsync(null, request).flux();
+
+        return collectionObs
+            .flatMap(documentCollectionResourceResponse -> {
+
+                DocumentCollection collection = documentCollectionResourceResponse.v;
+                if (collection == null) {
+                    throw new IllegalStateException("Collection cannot be null");
+                }
+
+                PartitionKeyDefinition pkDefinition = collection.getPartitionKey();
+                String pkSelector = createPkSelector(pkDefinition);
+                SqlQuerySpec querySpec = createLogicalPartitionScanQuerySpec(partitionKey, pkSelector);
+
+                Flux<Utils.ValueHolder<CollectionRoutingMap>> valueHolderMono = this.partitionKeyRangeCache
+                    .tryLookupAsync(BridgeInternal.getMetaDataDiagnosticContext(request.requestContext.cosmosDiagnostics),
+                        collection.getResourceId(),
+                        null,
+                        null).flux();
+                return valueHolderMono.flatMap(collectionRoutingMapValueHolder -> {
+
+                    CollectionRoutingMap routingMap = collectionRoutingMapValueHolder.v;
+                    if (routingMap == null) {
+                        throw new IllegalStateException("Failed to get routing map.");
+                    }
+
+                    String effectivePartitionKeyString = PartitionKeyInternalHelper
+                        .getEffectivePartitionKeyString(
+                            BridgeInternal.getPartitionKeyInternal(partitionKey),
+                            pkDefinition);
+
+                    //use routing map to find the partitionKeyRangeId of each
+                    // effectivePartitionKey
+                    PartitionKeyRange range =
+                        routingMap.getRangeByEffectivePartitionKey(effectivePartitionKeyString);
+
+                    final CosmosQueryRequestOptions effectiveOptions =
+                        ModelBridgeInternal.partitionKeyRangeIdInternal(
+                            ModelBridgeInternal.createQueryRequestOptions(options),
+                            range.getId());
+
+                    return createQuery(collectionLink, querySpec, effectiveOptions, Document.class, ResourceType.Document);
```

Comment:
partitionSplit handling. 
DocumentProducer which is used internally for cross-partition query is recursively split-proof.

Please verify two things:
1) that this code internally uses DocumentProducer (and hence benefit from DB being split proof.)
2) please make sure if partition split happens before DocumentProducer is instantiated still this is split proof.

Reply:
Fixed it - basically I had to change the code to include the PkRangeIdLookup within the retry policy as suspected. Thanks for catching this!
The Producers are used via the normal QueryexecutionContext factory - so we are good there.

Reply:
Actually I only moved the InvalidPartitionExceptionRetryPolicy up one level - so the change didn't impact partition splits. For partition splits my assumption is right now that readAllDocuments will behave like normal queries.

For queries the relevant PkRangeIds are retrieved form the query plan. The query plan is retrieved from the gateway before creating the QueryExecutionContext which then creates the DocumentProducer. QueryExecutionContext.executeAsync and DocumentProducer are handling partition splits.

For readAllDocuments the PkRangeId is set on CosmosQueryRequestOptions and this value is used by the QueryExecutionContext to create the DocumentProducer etc. Everything else is running through the same pipeline

So if there is a problem, because the pkValue -> pkRangeId resolution happens before creating the QueryExecutionContext/DocumentProducer in readAllDocuments the same problem exists for normal queries because the queryPlan has been also retrieved before creating ExecutionContext/DocumentProducer.

Worth some additional investigation and testing - but Mo and I decided offline that it is ok to do that after releasing the initial Beta version.